kernel/src/callback.rs

```diff
@@ -232,54 +283,70 @@ impl ProcessCallback {
+    ///
+    /// Otherwise, this function returns `true`.
     fn schedule(&mut self, r0: usize, r1: usize, r2: usize) -> bool {
-        let res = self
-            .app_id
+        self.app_id
             .kernel
             .process_map_or(false, self.app_id, |process| {
-                process.enqueue_task(process::Task::FunctionCall(process::FunctionCall {
-                    source: process::FunctionCallSource::Driver(self.callback_id),
-                    argument0: r0,
-                    argument1: r1,
-                    argument2: r2,
-                    argument3: self.appdata,
-                    pc: self.fn_ptr.as_ptr() as usize,
-                }))
-            });
-        if config::CONFIG.trace_syscalls {
-            debug!(
-                "[{:?}] schedule[{:#x}:{}] @{:#x}({:#x}, {:#x}, {:#x}, {:#x}) = {}",
-                self.app_id,
-                self.callback_id.driver_num,
-                self.callback_id.subscribe_num,
-                self.fn_ptr.as_ptr() as usize,
-                r0,
-                r1,
-                r2,
-                self.appdata,
-                res
-            );
-        }
-        res
-    }
+                if let Some(ptr) = self.fn_ptr {
+                    // This is a process callback pointing to a valid
+                    // address, try to schedule the callback
+                    let task_res =
+                        process.enqueue_task(process::Task::FunctionCall(process::FunctionCall {
+                            source: process::FunctionCallSource::Driver(self.callback_id),
+                            argument0: r0,
+                            argument1: r1,
+                            argument2: r2,
+                            argument3: self.appdata,
+                            pc: ptr.as_ptr() as usize,
+                        }));
 
-    // Required for the Tock 2.0 syscall interface
-    // pub(crate) fn function_pointer(&self) -> *const u8 {
-    //     self.fn_ptr.as_ptr() as *const u8
-    // }
-    //
-    // pub(crate) fn appdata(&self) -> u32 {
-    //     self.appdata as u32
-    // }
+                    if config::CONFIG.trace_syscalls {
+                        debug!(
+                            "[{:?}] schedule[{:#x}:{}] @{:#x}({:#x}, {:#x}, {:#x}, {:#x}) = {}",
+                            self.app_id,
+                            self.callback_id.driver_num,
+                            self.callback_id.subscribe_num,
+                            ptr.as_ptr() as usize,
+                            r0,
+                            r1,
+                            r2,
+                            self.appdata,
+                            task_res,
+                        );
+                    }
+
+                    task_res
+                } else {
+                    // This is a null callback, behave as if the
+                    // callback was scheduled
+                    true
+                }
```

Comment:
not printing that a null callback was scheduled when trace_syscalls is enabled seems like it could lead to confusion when a userspace bug leads to accidentally passing a null callback

Reply:
Agreed. We should introduce a trace message there as well.

Reply:
I've added an explicit message that a null callback was (not) scheduled.